File: Homework1/Hw1/Parser.cs
namespace Hw1;

public static class Parser
{
    public static void ParseCalcArguments(string[] args, 
        out double val1, 
        out CalculatorOperation operation, 
        out double val2)
    {
        val1 = val2 = 0;
        operation = CalculatorOperation.Undefined;

        if (!IsArgLengthSupported(args))
            throw new ArgumentException("Incorrect input");

        if (double.TryParse(args[0], out var value1) && double.TryParse(args[2], out var value2))
        {
            val1 = value1;
            val2 = value2;
            operation = ParseOperation(args[1]);
        }
        else
        {
            throw new ArgumentException("Incorrect input"); 
        }
    }

    private static bool IsArgLengthSupported(string[] args) => args.Length == 3;

    private static CalculatorOperation ParseOperation(string arg)
    {
        return arg switch
        {
            "+" => CalculatorOperation.Plus,
            "-" => CalculatorOperation.Minus,
            "*" => CalculatorOperation.Multiply,
            "/" => CalculatorOperation.Divide,
            _ => throw new InvalidOperationException("Invalid operation!")
        };
    }
}

File: Homework4/Hw4/Parser.fs
module Hw4.Parser

open System
open Hw4.Calculator


type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) =
    Array.length args = 3

let parseOperation (arg : string) =
    match arg with
    | "+" -> CalculatorOperation.Plus
    | "-" -> CalculatorOperation.Minus
    | "*" -> CalculatorOperation.Multiply
    | "/" -> CalculatorOperation.Divide
    | _ -> raise (System.ArgumentException("Incorrect Input!"))
    
let parseCalcArguments (args : string[]) =
    if not (isArgLengthSupported args) then
        raise (System.ArgumentException("Incorrect Input!"))

    let mutable arg1 = 0.0
    let mutable arg2 = 0.0

    match Double.TryParse(args.[0], &arg1), Double.TryParse(args.[2], &arg2) with
    | true, true ->
        let operation = parseOperation args.[1]
        { arg1 = arg1; operation = operation; arg2 = arg2 }
    | _ ->
        raise (System.ArgumentException("Incorrect Input!"))


